[DO NOT MERGE] Feature/bbnc 64 web ai sdlc test

cheshirsky opened the pull request:
PR represents engineer following  the [playbook](https://github.com/Backbase/golden-sample-app/blob/feature/BBNC-64-ai-sdlc-web/docs/playbooks/JIRA-001-playbook.md) on generating code. The playbook describes the following algorythm:
<img width="720" height="482" alt="image (2)" src="https://github.com/user-attachments/assets/9741540f-52a8-44f2-99d5-960172acbfd1" />


Relevant files are:
1. All source code (tests + actual code)
2. `docs/specs` folder (which is "part 1: specs" intermediate artefacts)
3. `docs/validation` folder (represents is "part 3" intermediate artefacts)

cheshirsky committed:
add ADRs

cheshirsky committed:
add test user story

cheshirsky committed:
add angular agent

cheshirsky committed:
add guideline

cheshirsky committed:
update playbook

cheshirsky committed:
update after 1st iteration

cheshirsky committed:
update it

cheshirsky committed:
update playbook

cheshirsky committed:
make sotry more realistic

cheshirsky committed:
update playbook

cheshirsky committed:
feedback

cheshirsky committed:
update UT instruction

cheshirsky committed:
update UT instruction

cheshirsky committed:
UT fix

cheshirsky committed:
specs(JIRA-001): approved spec and plan

cheshirsky committed:
S1 from JIRA-001

cheshirsky committed:
S2 from JIRA-001

cheshirsky committed:
S3 from JIRA-001

cheshirsky committed:
S4 add selection and routing

cheshirsky committed:
S5 add components display

cheshirsky committed:
3.2 code review

cheshirsky committed:
3.3 architecture review

cheshirsky committed:
3.4 product review

cheshirsky committed:
update playbook

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -51,20 +72,46 @@
+        </div>
+
+        <!-- Transactions List -->
+        <ul *ngIf="items.length > 0" class="bb-list">
```
We should prefer control flow in our templates as structural directives like ngIf and ngFor are deprecated. This would be particularly effective for the loop in this template, where you can use [at]empty as well.

Though appreciate this code change is inkeeping with existing template style, so this could be deferred if you choose.

cheshirsky replied:
+1. Right now old directives were used as LLM is forced to "follow repo standards", but we can influence it via ADRs

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -51,20 +72,46 @@
+              @@transactions.empty-state.title"
+            subtitle="There are no transactions for this account."
+            i18n-subtitle="
+              Transactions empty state subtitle - 'There are no transactions for
```
Its unconventional to include the actual default sentence in the i18n key description

cheshirsky replied:
Agreed, I'll check if it can be addressed via adding it to the ADR

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.ts:
```diff
@@ -71,12 +101,47 @@ export class TransactionsViewComponent {
+   * Sets up default account selection logic.
+   * RULE: Navigates to first account if no account or invalid account in URL.
+   */
+  private initDefaultAccountSelection(): void {
```
private methods should be located after public / protected ones

cheshirsky replied:
Added to the todo

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+          (importedModule as { name?: string })?.name === 'AccountSelectorModule'
+      );
+
+      // Assert
```
These inline comments on test case structure are redundant

cheshirsky replied:
Issue caused by instructions to use reasononing as inline comments to pivot code generation, prompt needs hardening. Added to my todo

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+    });
+
+    // Edge case: Module exports are available after import
+    it('should_export_components_when_module_is_imported', () => {
```
Poor formatting on test descriptions. There is no need for these to be underscored.

cheshirsky replied:
Yeh, I am missing ut agent entirely on this prototype, as my point was to prove that tests improve code.

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.ts:
```diff
@@ -71,12 +101,47 @@ export class TransactionsViewComponent {
+      )
+      .subscribe(({ selectedAccount, accounts }) => {
+        // RULE: Navigate to first account if no valid selection
+        if (!selectedAccount) {
```
Having conditions inside a subscription handler is often a smell. Consider using filter operator to apply this logic instead (filter(({ selectedAccount }) => !selectedAccount))

cheshirsky replied:
Added ADR to the list of missing artefacts

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -11,6 +11,27 @@
     </div>
 
     <div class="bb-card__body">
+      <!-- Account Selector -->
```
This is subjective, but some of the inline commenting in this template is unhelpful in my view. Here we have a comment which is repeating what the well-named component selector is already communicating.

daiscog replied:
I agree.  Most code comments are prone to becoming stale over time, so become a distraction.  If code is well-written, explanatory comments aren't needed.

cheshirsky replied:
yeh, leaked LCPs, sorry :)

JonnyHeavey commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.ts:
```diff
@@ -31,6 +32,35 @@ export class TransactionsViewComponent {
     map((params) => params.get('account'))
   );
 
+  /**
```
This tsdoc describes the implementation (which itself is clear) rather than what the property itself is

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -11,6 +11,27 @@
+      <!-- Account Selector -->
+      <div class="bb-block bb-block--md">
+        <bb-account-selector-ui
+          data-role="transactions-view__account-selector"
```
We recommend using accessibility (role-based) selectors in our tests over `data-role` attributes, with `data-testid` only as a fallback when a role-based selector is too cumbersome.

cheshirsky replied:
added to the todo

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -51,20 +72,46 @@
-          *ngFor="let item of items | filterTransactions : filter"
-          class="bb-list__item"
-          data-role="transactions-view__item-container"
+      <ng-container *ngIf="transactions$ | async as items; else loading">
```
`*ngIf` is deprecated.  We should be using `@if`.

cheshirsky replied:
added

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.html:
```diff
@@ -51,20 +72,46 @@
+            class="bb-list__item"
+            data-role="transactions-view__item-container"
+          >
+            <a class="list-link" [routerLink]="['./', item.id]">
```
Nit-pick, but I'm pretty sure routerLink here can be simplified to `[routerLink]="item.id"` since the routerLink directive is relative to the current activated route by default

cheshirsky replied:
not sure if I need to reflect it anywhere, as it comes from the repo itself. Probably ADRs always prioritiesd over repo-level stuff.

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.ts:
```diff
@@ -71,12 +101,47 @@ export class TransactionsViewComponent {
     private readonly router: Router,
     private readonly transactionsService: TransactionsHttpService,
     private readonly arrangementsService: ArrangementsService,
+    private readonly destroyRef: DestroyRef,
```
We should prefer `inject` over constructor injection

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.ts:
```diff
@@ -71,12 +101,47 @@ export class TransactionsViewComponent {
+   * RULE: Navigates to first account if no account or invalid account in URL.
+   */
+  private initDefaultAccountSelection(): void {
+    combineLatest({
```
This feels like a very convoluted way of selecting the first account by default, and also adds a navigation to the history stack (I'd be suspicious about what effect using the browser back button would have).

A much simpler approach would be for the `selectedAccount$` observable to emit the first account if the `accountId$` observable emits undefined (or even any invalid value) then we won't need to have this programmatic navigation to add the first account id to the query params

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.spec.ts:
```diff
@@ -195,4 +298,652 @@ describe('TransactionsViewComponent', () => {
+
+      // Assert
+      expect(accounts[0].number).toBe('123456789');
+    });
```
These tests could be collapsed into one test that does 

```ts
expect(accounts[0]).toStrictEqual({
  // ... all of the properties here
})
```

cheshirsky replied:
missing UT agent. Added

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.spec.ts:
```diff
@@ -195,4 +298,652 @@ describe('TransactionsViewComponent', () => {
+      (fixture.componentInstance as unknown as { accounts$: Observable<AccountSelectorItem[]> }).accounts$;
+
+    // Happy path: accounts$ emits mapped arrangement data
+    it('should_emit_mapped_accounts_when_arrangements_are_loaded', async () => {
```
I'm not a fan of the underscores instead of spaces in the test names - is there a reason for these?

cheshirsky replied:
agreed, naming should be part of UT agent clearly

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.spec.ts:
```diff
@@ -195,4 +298,652 @@ describe('TransactionsViewComponent', () => {
+     * Uses type assertion since property is being added in S2.
+     */
+    const getAccounts$ = () =>
+      (fixture.componentInstance as unknown as { accounts$: Observable<AccountSelectorItem[]> }).accounts$;
```
Are these "helpers" still needed? They could mask genuine typing errors, and the JSDoc description sounds stale.

cheshirsky replied:
Is it a smell in your opinion, or style preference?

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.component.spec.ts:
```diff
@@ -195,4 +298,652 @@ describe('TransactionsViewComponent', () => {
+    });
+  });
+
+  describe('selectedAccount$ observable (S2)', () => {
```
Just a thought: For non-trivial observables, I like to use marbles testing so we can assert how an observable will change over time, in reaction to new values being emitted by the source observables.

cheshirsky replied:
Added

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+  ɵmod?: IvyModuleDef;
+  ɵinj?: IvyModuleDef;
+}
+
```
Why would we care about Angular's internal APIs?

cheshirsky replied:
mocking for the module test

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+
+      // Act & Assert
+      expect(moduleSetup).not.toThrow();
+    });
```
This feels like a really low-value test.  If using the module throws then all other tests that use it will fail.

cheshirsky replied:
Good point

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+        (importedModule) =>
+          importedModule === AccountSelectorModule ||
+          (importedModule as { name?: string })?.name === 'AccountSelectorModule'
+      );
```
I'm not sure I see the point of this test either.  We should be testing business logic and behaviours not Angular boilerplate code.

We never need to assert that one module imports another  - the Angular compilation will fail if we have missing module imports that our components need.

(Same for all the other similar tests)

cheshirsky replied:
fair, added to agent

daiscog commented on libs/transactions-journey/internal/feature-transaction-view/src/lib/components/transactions-view/transactions-view.module.spec.ts:
```diff
@@ -0,0 +1,122 @@
+
+      // Assert
+      expect(module).toBeDefined();
+    });
```
Again, feels like we're testing Angular constructs here, not our own code.  Tbh, the vast majority of `.module.ts` files do not need any unit tests at all, as they are 100% Angular boilerplate.  If it had things like `.forRoot` or other functions in the class that contained our own logic, or if it set up complicated providers, then it's worth adding tests for those.  But this entire file feels like a box-ticking exercise.

cheshirsky replied:
added

JonnyHeavey commented:
Overall, this is a reasonable start for an automated contribution, with the core of the implementation providing the required functionality. However, whilst the majority of the issues raised here are of a stylistic nature, they should be taken seriously to avoid degradation of codebase maintainability.

I'm sure with additional context (and perhaps automated reivew/verification loops) we can steer the agent to more preferable codestyle and patterns.

cheshirsky committed:
add notes on future improvements

cheshirsky commented:
@daiscog @JonnyHeavey added the list of outstanding issues with resolutions for them, WDYT?

cheshirsky committed:
add the playbook

cheshirsky committed:
update playbook, add issues

cheshirsky committed:
update ADRs

cheshirsky committed:
add skills and behaviours

cheshirsky committed:
add agents

cheshirsky committed:
use agents

cheshirsky committed:
clean up experiment

cheshirsky committed:
Revert "S5 add components display"

This reverts commit 8b91e6eda99fe674434bb4c6b33fda7688312ab8.

cheshirsky committed:
Revert "S4 add selection and routing"

This reverts commit 2d854c352e7087ff2300a8e6332eb9c3e7a2a234.

cheshirsky committed:
Revert "S3 from JIRA-001"

This reverts commit bfe94a079825481eca5cbc86c06024ebd7c99b6d.

cheshirsky committed:
Revert "S2 from JIRA-001"

This reverts commit f4ddebf47e043726b3113a24b32f350b2b2f7002.

cheshirsky committed:
Revert "S1 from JIRA-001"

This reverts commit 933d4e008d56abc1119fa5770df2b55ce1e1257c.

cheshirsky committed:
specs(JIRA-001): approved spec and plan

cheshirsky committed:
feat(JIRA-001): step 1 - add EntitlementsGuard to transactions route

cheshirsky committed:
feat(JIRA-001): step 2 - import AccountSelectorModule

cheshirsky committed:
feat(JIRA-001): step 3 - add account selector logic with TDD

cheshirsky committed:
feat(JIRA-001): step 4 - add account selector template with i18n

cheshirsky committed:
feat(JIRA-001): step 5 - add empty state for no transactions

cheshirsky committed:
chore(JIRA-001): add code review validation

cheshirsky committed:
fix(JIRA-001): use correct bb-account-selector-ui API

cheshirsky committed:
docs: add UI component API verification to playbooks

Lesson learned: Do not assume UI component APIs from naming conventions.
Always verify inputs/outputs against existing codebase usage.

cheshirsky committed:
chore(JIRA-001): add architecture and product reviews

cheshirsky committed:
update ignore files

cheshirsky committed:
upd docs

cheshirsky committed:
Revert "chore(JIRA-001): add architecture and product reviews"

This reverts commit 422add21fde165c2a8410ec72b57131a32a0d3b8.

cheshirsky committed:
Revert "fix(JIRA-001): use correct bb-account-selector-ui API"

This reverts commit 34cd8f9aaa5560a28e54ad2637f2102190383ce3.

cheshirsky committed:
Revert "chore(JIRA-001): add code review validation"

This reverts commit 8e9695671290b5202f1c005ce2215ad1a9d829ef.

cheshirsky committed:
Revert "feat(JIRA-001): step 5 - add empty state for no transactions"

This reverts commit 79d8cb7ebbbaab64922a5f351b593e648d40d1ff.

cheshirsky committed:
Revert "feat(JIRA-001): step 4 - add account selector template with i18n"

This reverts commit 5e89646626aea6d8ac3bbb22c27d8576b54f3220.

cheshirsky committed:
Revert "feat(JIRA-001): step 3 - add account selector logic with TDD"

This reverts commit 0073bf1011fdda1853050f57b05b2e30cf90fe4a.

cheshirsky committed:
Revert "feat(JIRA-001): step 2 - import AccountSelectorModule"

This reverts commit 4774cec111024c46a76d9016ade352b8af8bbf3d.

cheshirsky committed:
Revert "feat(JIRA-001): step 1 - add EntitlementsGuard to transactions route"

This reverts commit 63d277a00868457a1c3a21a18945beb3a8362c05.

cheshirsky committed:
Revert "specs(JIRA-001): approved spec and plan"

This reverts commit 82388f8e2f627918025a5d3049dbff68371e0c27.

cheshirsky committed:
specs(JIRA-001): approved spec, solution design, and execution plan

cheshirsky committed:
feat(JIRA-001): step 1 - import AccountSelectorModule

cheshirsky committed:
feat(JIRA-001): step 2 - add account selector to template

cheshirsky committed:
feat(JIRA-001): step 3 - implement account selection with URL

cheshirsky committed:
feat(JIRA-001): step 4 - auto-select first account on load

cheshirsky committed:
feat(JIRA-001): step 5 - add empty transactions state

cheshirsky committed:
docs(JIRA-001): add validation reports (code, architecture, product)

cheshirsky committed:
Revert "docs(JIRA-001): add validation reports (code, architecture, product)"

This reverts commit 56b84e27a267a33ffff7fca5ebcedbac7f00d112.

cheshirsky committed:
Revert "feat(JIRA-001): step 5 - add empty transactions state"

This reverts commit 2d3b246972b3a11d4973b321a5a7f659622a9585.

cheshirsky committed:
Revert "feat(JIRA-001): step 4 - auto-select first account on load"

This reverts commit 61ed0fbf2671685ecd1e93c7bcda8bae48af32f8.

cheshirsky committed:
Revert "feat(JIRA-001): step 3 - implement account selection with URL"

This reverts commit d282593f3905b0a8fa72d559be853ff4ee5df487.

cheshirsky committed:
Revert "feat(JIRA-001): step 2 - add account selector to template"

This reverts commit fae784d40f8c6c1a55a7d318eee9843e49bc5cc0.

cheshirsky committed:
Revert "feat(JIRA-001): step 1 - import AccountSelectorModule"

This reverts commit 298d15744bb91d89ae10a39a20a338cf84f1cdee.

cheshirsky committed:
Revert "specs(JIRA-001): approved spec, solution design, and execution plan"

This reverts commit 138066d439a65cea576670c6040d47af53395b5c.

cheshirsky committed:
specs(JIRA-001): approved spec and plan

cheshirsky committed:
feat(JIRA-001): step 1 - import AccountSelectorModule

cheshirsky committed:
feat(JIRA-001): step 2 - add account selector to template

cheshirsky committed:
feat(JIRA-001): step 3 - add account selection logic

cheshirsky committed:
feat(JIRA-001): step 4 - verify transaction filtering by account

cheshirsky committed:
feat(JIRA-001): step 5 - add EntitlementsGuard to transactions route

cheshirsky committed:
brush up structure

cheshirsky committed:
clean up

cheshirsky committed:
redo design with the new spec

cheshirsky committed:
redo validation

cheshirsky committed:
harden SI validation

cheshirsky committed:
minor fixes
